### DISCUSSION TITLE
HA setup in Strongswan 5.9 on RHEL8

### DISCUSSION THREAD

**Discussion by AshnaJindal01** [strongswan/strongswan#1116] [General]
I am trying to setup HA solution for Software tunnels based on Strongswan 5.9 and the base OS is RHEL8.x

I am facing issues while using CLUSTERIP linux module on RHEL8 as this seems to be deprecated. I tried to use "Cluster" module instead but even that is giving some errors.

I have been following below Strongswan documentation for HA setup:
https://docs.strongswan.org/docs/5.9/features/highAvailability.html#:~:text=As%20an%20alternative%20strongSwan%20provides,handled%20by%20a%20different%20node.

ip address add 192.168.0.200/24 dev eth0
iptables -A INPUT -i eth0 -d 192.168.0.200 -j CLUSTERIP --new \
   --hashmode sourceip --clustermac 01:00:5e:00:00:20 \
   --total-nodes 2 --local-node 0

The above step is not working and while using "cluster" option, it does not have any parameter as clustermac to pass due to which we cannot set the rule for destination IP to jump to the new target having multi MAC address as 01:00:5e:00:00:20.

Can someone please help me in setting up the HA ?

**Comment by Thermi** [comment]
What ARE the errors?

**Reply by Thermi** [reply]
I can already tell you (I once looked at rewriting that tutorial for using CLUSTER instead of CLUSTERIP) that at least sending of ARP replies with the cluster MAC for ARP requests for the cluster IP is missing (probably needs a proxy arp record and enabling of proxy arp, or a userspace daemon that takes care of that, same for ndp if IPv6 is used).
That's just from the top of my hat. To figure out the details you'd need to look at the source code of the CLUSTERIP kernel module to figure out what it all does.

**Reply by twogee** [reply]
`cluster` [documentation](https://ipset.netfilter.org/iptables-extensions.man.html#lbAJ) suggests using `arptables`. Would it suffice on newer Linux distributions with NetFilter-based iptables? Currently, trying to use `CLUSTERIP` on Ubuntu 22.04 results in
```
iptables v1.8.7 (nf_tables): RULE_APPEND failed (Operation not supported): rule in chain INPUT

kernel: ipt_CLUSTERIP: ClusterIP Version 0.8 loaded successfully
kernel: ipt_CLUSTERIP: cannot use CLUSTERIP from nftables compat
```
or
```
iptables: Invalid argument. Run ’dmesg’ for more information.

kernel: bpfilter: loaded bpfilter_umh pid …
kernel: ipt_CLUSTERIP: bad local_nodes[0] 0
```

**Comment by AshnaJindal01** [comment]
I have created a new chain in iptables with name "S_TUNNEL" and has  tried to divert the traffic, hitting strongswan tunnel gateway to this chain using Linux modprobe "cluster" , as this is the only ipt_cluster module available in RHEL8.


$iptables -L -n
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
S_TUNNEL   all  --  0.0.0.0/0            192.168.0.200        cluster node_mask=0x00000001 total_nodes=2 hash_seed=0x00000000

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain CLUSTERIP (0 references)
target     prot opt source               destination

Chain S_TUNNEL (1 references)
target     prot opt source               destination


==================================================================
When trying to attach the above with CLuster external multi MAC IP address, the error is as below 👍 

iptables -A INPUT -d 192.168.0.200/32  -j S_TUNNEL -m cluster --cluster-total-nodes 2 --cluster-local-node 1 --cluster-hash-seed 0 --clustermac 01:00:5e:00:01:01
iptables v1.8.4 (nf_tables): unknown option "--clustermac"

**Reply by Thermi** [reply]
Please provide the output of `iptables-save` instead of `iptables -L`. Also, the cluster module is NOT a drop-in replacements for CLUSTERIP. They do different things! The man page for iptables-extensions that describes all iptables extensions lists the valid parameters. You can also ask iptables for the help messages for the listed modules and targets by just appending -h to the parameter list. E.g.: iptables -A INPUT -m cluster -h
